lib/traject/macros/custom.rb

```diff
@@ -1,9 +1,82 @@
+      #
+      # Split with em dash along v,x,y,z
+      def process_subject_hierarchy(fields)
+        lambda do |record, accumulator|
+          return nil unless record.is_a? MARC::Record
+
+          subjects = []
+          split_on_subfield = %w[v x y z]
+          Traject::MarcExtractor.cached(fields).collect_matching_lines(record) do |field, spec, extractor|
+            subject = extractor.collect_subfields(field, spec).first
+            unless subject.nil?
+              field.subfields.each do |s_field|
+                subject = subject.gsub(" #{s_field.value}", "#{SEPARATOR}#{s_field.value}") if split_on_subfield.include?(s_field.code)
+              end
+              subject = subject.split(SEPARATOR)
+              subject = subject.map { |s| Traject::Macros::Marc21.trim_punctuation(s) }.join(SEPARATOR)
+              subjects << subject # if include_subject
+            end
+          end
+
+          accumulator.replace(subjects.compact.uniq)
+        end
+      end
+
+      # For the split subject facet
+      #
+      # Split with em dash along v,x,y,z
+      def process_subject_topic_facet(fields)
+        lambda do |record, accumulator|
+          return nil unless record.is_a? MARC::Record
+
+          subjects = []
+          split_on_subfield = %w[v x y z]
+          Traject::MarcExtractor.cached(fields).collect_matching_lines(record) do |field, spec, extractor|
+            subject = extractor.collect_subfields(field, spec).first
+            unless subject.nil?
+              field.subfields.each do |s_field|
+                subject = subject.gsub(" #{s_field.value}", "#{SEPARATOR}#{s_field.value}") if split_on_subfield.include?(s_field.code)
+              end
+              subject = subject.split(SEPARATOR)
+              subjects << subject.map { |s| Traject::Macros::Marc21.trim_punctuation(s) }
+            end
+          end
+
+          accumulator.replace(subjects.flatten.compact.uniq)
+        end
+      end
+
+      # For genre facet and display
+      #
+      # limit to subfield $2 vocabularies for 655|*7 genres
+      def process_genre(fields)
+        lambda do |record, accumulator|
+          return nil unless record.is_a? MARC::Record
+
+          genres = []
+          vocabulary = %w[lcgft fast]
+          Traject::MarcExtractor.cached(fields).collect_matching_lines(record) do |field, spec, extractor|
+            genre = extractor.collect_subfields(field, spec).first
+            include_genre = true
+            unless genre.nil?
+              include_genre = vocabulary.include?(field['2'].to_s.downcase) if (field.tag == '655') && (field.indicator2 == '7')
+              genres << Traject::Macros::Marc21.trim_punctuation(genre) if include_genre
+            end
+          end
+
+          accumulator.replace(genres).uniq!
+        end
+      end
+
       NOT_FULLTEXT = /addendum|appendices|appendix|appendixes|cover|excerpt|executive summary|index/i.freeze
```

Comment:
I wonder if this constant should be put up on line 3 🤔 just for consistency

Reply:
(I know this is unrelated)

Reply:
ahaa yeah makes sense